# Basics of MLP
- Objective: create vanilla neural networks (i.e., Multilayer perceptrons) for simple regression/classification tasks with Keras

## MLP Structures
- Each MLP model is consisted of one input layer, several hidden layers, and one output layer
- Number of neurons in each layer is not limited
<img src="http://cs231n.github.io/assets/nn1/neural_net.jpeg" style="width: 300px"/>
<br>
<center>**MLP with one hidden layer**</center>
- Number of input neurons: 3
- Number of hidden neurons: 4
- Number of output neurons: 2


<img src="http://cs231n.github.io/assets/nn1/neural_net2.jpeg" style="width: 500px"/>
<br>
<center>**MLP with two hidden layers**</center>
- Number of input neurons: 3
- Number of hidden neurons: (4, 4)
- Number of output neurons: 1


## MLP for Regression tasks
- When the target (**y**) is continuous (real)
- For loss function and evaluation metric, mean squared error (MSE) is commonly used

In [ ]:
from keras.datasets import boston_housing
(x_train, y_train),(x_test,y_test) = boston_housing.load_data()

65536/57026 [==================================] - 0s 0us/step


In [ ]:
len(x_train)

404

### Dataset Description
- Boston housing dataset has total 506 data instances (404 training & 102 test)
- 13 attributes (features) to predict "the median values of the houses at a location"
- Doc: https://keras.io/datasets/

In [ ]:
print(x_train.shape)
print(x_test.shape)

(404, 13)
(102, 13)


### 1. Creating a model
- Keras model object can be created with Sequential class
- At the outset, the model is empty per se. It is completed by **'adding'** additional layers and compilation
- Doc: https://keras.io/models/sequential/

In [ ]:
from keras.models import Sequential

In [ ]:
model = Sequential()

### 1-1. Adding layers
- Keras layers can be **added** to the model
- Adding layers are like stacking lego blocks one by one
- Doc: https://keras.io/layers/core/

In [ ]:
from keras.layers import Dense, Activation

In [ ]:
model.add(Dense(10,input_shape = (13,),activation='sigmoid'))
model.add(Dense(10,activation='sigmoid'))
model.add(Dense(10,activation='sigmoid'))
model.add(Dense(1))

### 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
    - Doc (optimizers): https://keras.io/optimizers/
    - Doc (losses): https://keras.io/losses/

In [ ]:
import tensorflow as tf
from keras import optimizers
gd = tf.keras.optimizers.SGD(learning_rate = 0.01)

In [ ]:
model.compile(optimizer = gd,loss ='mean_squared_error',metrics = ['mse'])

### Summary of the model

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 10)                140       
                                                                 
 dense_1 (Dense)             (None, 10)                110       
                                                                 
 dense_2 (Dense)             (None, 10)                110       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 371
Trainable params: 371
Non-trainable params: 0
_________________________________________________________________


### 2. Training
- Training the model with training data provided

In [ ]:
model.fit(x_train,y_train, batch_size = 32, epochs = 100,verbose = 1,validation_split = 0.20)

Epoch 1/100
11/11 [==============================] - 0s 9ms/step - loss: 84.5700 - mse: 84.5700 - val_loss: 93.3662 - val_mse: 93.3662
Epoch 2/100
11/11 [==============================] - 0s 4ms/step - loss: 84.9189 - mse: 84.9189 - val_loss: 88.2852 - val_mse: 88.2852
Epoch 3/100
11/11 [==============================] - 0s 6ms/step - loss: 84.3896 - mse: 84.3896 - val_loss: 88.9679 - val_mse: 88.9679
Epoch 4/100
11/11 [==============================] - 0s 6ms/step - loss: 84.5675 - mse: 84.5675 - val_loss: 89.7445 - val_mse: 89.7445
Epoch 5/100
11/11 [==============================] - 0s 6ms/step - loss: 84.5642 - mse: 84.5642 - val_loss: 93.2356 - val_mse: 93.2356
Epoch 6/100
11/11 [==============================] - 0s 5ms/step - loss: 85.1746 - mse: 85.1746 - val_loss: 84.6271 - val_mse: 84.6271
Epoch 7/100
11/11 [==============================] - 0s 6ms/step - loss: 84.9881 - mse: 84.9881 - val_loss: 86.9963 - val_mse: 86.9963
Epoch 8/100
11/11 [==============================] - 0s

### 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
    - Doc (metrics): https://keras.io/metrics/

In [ ]:
results = model.evaluate(x_test,y_test)
print(model.metrics_names)
print(results)

4/4 [==============================] - 0s 3ms/step - loss: 83.2682 - mse: 83.2682
['loss', 'mse']
[83.26820373535156, 83.26820373535156]


## MLP for classification tasks
- When the target (**y**) is discrete (categorical)
- For loss function, cross-entropy is used and for evaluation metric, accuracy is commonly used

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

In [ ]:
data = load_breast_cancer()

In [39]:
data

{'DESCR': '.. _breast_cancer_dataset:\n\nBreast cancer wisconsin (diagnostic) dataset\n--------------------------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry\n        - fractal dimension ("coastline approximation" - 1)\n\n        The mean, standard error, and "worst" or largest (mean of the three\n        worst/largest values) of these features were computed for each image,\n        resulting in 30 features.

In [ ]:
y = data.target

In [ ]:
x = data.data

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.25,random_state = 60)

### Dataset Description
- Breast cancer dataset has total 569 data instances (212 malign, 357 benign instances)
- 30 attributes (features) to predict the binary class (M/B)
- Doc: http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer

In [30]:
print(x_train.shape)
print(x_test.shape)

(426, 30)
(143, 30)


### 1. Creating a model
- Same with regression model at the outset

In [31]:
from keras.models import Sequential

In [36]:
model = Sequential()

### 1-1. Adding layers
- Keras layers can be **added** to the model
- Adding layers are like stacking lego blocks one by one
- It should be noted that as this is a classification problem, sigmoid layer (softmax for multi-class problems) should be added
- Doc: https://keras.io/layers/core/

In [37]:
model.add(Dense(10,input_shape = (30,)))
model.add(Activation('sigmoid'))
model.add(Dense(10))
model.add(Activation('sigmoid'))
model.add(Dense(10))
model.add(Activation('sigmoid'))
model.add(Dense(1))
model.add(Activation('sigmoid'))

In [ ]:
"""#the above cell is equivalent to 
model.add(Dense(10,input_shape = (30,),activation='sigmoid'))
model.add(Dense(10,activation='sigmoid'))
model.add(Dense(10,activation='sigmoid'))
model.add(Dense(1))"""

### 1-2. Model compile
- Keras model should be "compiled" prior to training
- Types of loss (function) and optimizer should be designated
    - Doc (optimizers): https://keras.io/optimizers/
    - Doc (losses): https://keras.io/losses/

In [38]:
import tensorflow as tf
from keras import optimizers
gd = tf.keras.optimizers.SGD(learning_rate = 0.01)

In [40]:
model.compile(optimizer = gd,loss ='binary_crossentropy',metrics = ['accuracy'])

### Summary of the model

In [41]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 10)                310       
                                                                 
 activation_8 (Activation)   (None, 10)                0         
                                                                 
 dense_13 (Dense)            (None, 10)                110       
                                                                 
 activation_9 (Activation)   (None, 10)                0         
                                                                 
 dense_14 (Dense)            (None, 10)                110       
                                                                 
 activation_10 (Activation)  (None, 10)                0         
                                                                 
 dense_15 (Dense)            (None, 1)                

### 2. Training
- Training the model with training data provided

In [ ]:
model.fit(x_train,y_train, batch_size = 32, epochs = 100,verbose = 1,validation_split = 0.20)

### 3. Evaluation
- Keras model can be evaluated with evaluate() function
- Evaluation results are contained in a list
    - Doc (metrics): https://keras.io/metrics/

In [43]:
results = model.evaluate(x_test,y_test)
print(model.metrics_names)
print(results)

5/5 [==============================] - 0s 4ms/step - loss: 0.6514 - accuracy: 0.6434
['loss', 'accuracy']
[0.6514446139335632, 0.6433566212654114]
